### ISSUE TITLE
Can't use setgid

### ISSUE THREAD

**Issue by markc** [lxc/lxd#4294]
# Required information

 * Distribution: Ubuntu + kubuntu-CI
 * Distribution version: 18.04
 * The output of "lxc info" or if that fails:
```
config:
  core.https_address: 192.168.0.2:8443
  core.trust_password: true
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses:
  - 192.168.0.2:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    [...]
    -----END CERTIFICATE-----
  certificate_fingerprint: XXXXXXX
  driver: lxc
  driver_version: 2.1.1
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.15.0-10-lowlatency
  server: lxd
  server_pid: 2953
  server_version: "2.21"
  storage: zfs
  storage_version: 0.7.5-1ubuntu1
```
# Issue description

All attempts to use setgid fail.

# Steps to reproduce

 1. Step one
Create a new ubuntu bionic container.

 2. Step two
Add postfix (as one example) which requires setgid on /usr/sbin/postdrop and /usr/sbin/postqueue and the mail server becomes unusable.

 3. Step three
Manually attempt to use chmod 02555 on the above 2 binaries silently fails.

# Information to attach

 - [ ] Any relevant kernel output (`dmesg`)
 - [ ] Container log (`lxc info NAME --show-log`)
[lxc-info-log.txt](https://github.com/lxc/lxd/files/1773515/lxc-info-log.txt)
 - [ ] Container configuration (`lxc config show NAME --expanded`)
[lxc-info-expanded.txt](https://github.com/lxc/lxd/files/1773514/lxc-info-expanded.txt)
 - [ ] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)
 [lxd.log.txt](https://github.com/lxc/lxd/files/1773513/lxd.log.txt)
- [ ] Output of the client with --debug
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monito

**Comment by stgraber** [comment]
That's very odd, I strongly suspect a zfs bug...

**Comment by markc** [comment]
Any suggestions how to test or debug this further? The host is ext4 but should I try a btrfs storage pool?

**Comment by stgraber** [comment]
```
root@c1:~# touch a
root@c1:~# chmod g+s a
root@c1:~# touch b
root@c1:~# chown 0:117 b
root@c1:~# chmod g+s b
root@c1:~# stat a
  File: a
  Size: 0         	Blocks: 1          IO Block: 131072 regular empty file
Device: 43h/67d	Inode: 33890       Links: 1
Access: (2644/-rw-r-Sr--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2018-03-02 03:32:47.019430367 +0000
Modify: 2018-03-02 03:32:47.019430367 +0000
Change: 2018-03-02 03:32:49.459445015 +0000
 Birth: -
root@c1:~# stat b
  File: b
  Size: 0         	Blocks: 1          IO Block: 131072 regular empty file
Device: 43h/67d	Inode: 34186       Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (  117/postdrop)
Access: 2018-03-02 03:32:50.907453706 +0000
Modify: 2018-03-02 03:32:50.907453706 +0000
Change: 2018-03-02 03:33:01.299516054 +0000
 Birth: -
root@c1:~# 
```

**Comment by stgraber** [comment]
```
root@c1:~# mkdir tmpfs
root@c1:~# mount -t tmpfs tmpfs tmpfs
root@c1:~# cd tmpfs/
root@c1:~/tmpfs# touch a
root@c1:~/tmpfs# chmod g+s a
root@c1:~/tmpfs# touch b
root@c1:~/tmpfs# chown 0:117 b
root@c1:~/tmpfs# chmod g+s b
root@c1:~/tmpfs# stat a
  File: a
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: 65h/101d	Inode: 3           Links: 1
Access: (2644/-rw-r-Sr--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2018-03-02 03:33:35.783722623 +0000
Modify: 2018-03-02 03:33:35.783722623 +0000
Change: 2018-03-02 03:33:40.507750883 +0000
 Birth: -
root@c1:~/tmpfs# stat b
  File: b
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: 65h/101d	Inode: 4           Links: 1
Access: (2644/-rw-r-Sr--)  Uid: (    0/    root)   Gid: (  117/postdrop)
Access: 2018-03-02 03:33:42.131760597 +0000
Modify: 2018-03-02 03:33:42.131760597 +0000
Change: 2018-03-02 03:33:46.227785091 +0000
 Birth: -
root@c1:~/tmpfs# 
```

**Comment by stgraber** [comment]
Right, so the first dump shows that setgid silently fails when doing it on a file that has a uid/gid set, it may be tied to whether the user is a member of the group or not (not sure).

**Comment by stgraber** [comment]
The second shows the exact same test on a tmpfs inside the same container, which works as expected.

**Comment by stgraber** [comment]
So it's not a container problem in the sense that it's not some seccomp/apparmor policy silently blocking the chown call but it's something weird going on inside of ZFS.

I'd recommend you file a critical bug against the Linux kernel in Ubuntu as that's where the zfs kernel module is. I suspect this is a regression in ZFS 0.7 which recently landed.

**Comment by stgraber** [comment]
If you do file that bug, please post the URL here so I can subscribe our team to it to track it.

@sforshee @ColinIanKing FYI ^

**Comment by markc** [comment]
FWIW I finally managed to change all my lxd hosts over to btrfs storage pools and I can absolutely confirm the same server setup now allows setgid with btrfs on 3.0.0.beta2. I'm not sure I'll have time to figure out how to file a bug with ubuntu as I need to catch up with a long overdue hosting system update. BTW I had to rsync everything around as copy would just not work, for me.

**Comment by stgraber** [comment]
Ok, I filed this on your behalf: https://launchpad.net/ubuntu/+source/linux/+filebug

**Comment by markc** [comment]
Just a FWIW, I  updated my bionic system and reinstalled lxd, switched back to zfs, and still have the same problem. I only mention this because the launchpad bug #1753288 says this setgid problem should be fixed in zfs-linux 0.7.5-1ubuntu5 but it doesn't appear to be so...
```
postfix/postfix-script: warning: not set-gid or not owner+group+world executable: /usr/sbin/postqueue
postfix/postfix-script: warning: not set-gid or not owner+group+world executable: /usr/sbin/postdrop
postdrop: warning: mail_queue_enter: create file maildrop/831097.21335: Permission denied
```
This is using lxd 3.0.0.beta5 and zfsutils-linux 0.7.5-1ubuntu7
```
  driver: lxc
  driver_version: 3.0.0.beta1
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.15.0-12-lowlatency
  server: lxd
  server_pid: 7007
  server_version: 3.0.0.beta5
  storage: zfs
  storage_version: 0.7.5-1ubuntu1
  server_clustered: false
  server_name: mbox
```
OIC, it seems lxd 3.0.0.beta5 is built against zfs 0.7.5-1ubuntu1. Hopefully the next update will be built against 0.7.5-1ubuntu5+. Like I say, just a FWIW.

**Comment by stgraber** [comment]
I believe the Launchpad bug also mentions that the kernel needs a matching update, which it is yet to receive.

The zfs source in Ubuntu is used for both the tools and for the kernel module, but as we ship pre-built kernel modules, you need to wait for the next kernel to land to have the fix.
